common/src/main/java/com/microsoft/identity/common/internal/telemetry/events/ApiStartEvent.java

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.telemetry.events;
+
+import android.support.annotation.NonNull;
+
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.request.AcquireTokenOperationParameters;
+import com.microsoft.identity.common.internal.request.AcquireTokenSilentOperationParameters;
+import com.microsoft.identity.common.internal.request.BrokerAcquireTokenOperationParameters;
+import com.microsoft.identity.common.internal.request.BrokerAcquireTokenSilentOperationParameters;
+import com.microsoft.identity.common.internal.request.OperationParameters;
+import com.microsoft.identity.common.internal.util.StringUtil;
+
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.NoSuchAlgorithmException;
+
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.*;
+
+public class ApiStartEvent extends BaseEvent {
+    private static final String TAG = ApiStartEvent.class.getSimpleName();
+
+    public ApiStartEvent() {
+        super();
+        names(TELEMETRY_EVENT_API_EVENT_START);
+    }
+
+    public ApiStartEvent putProperties(OperationParameters parameters) {
+        this.put(TELEMETRY_KEY_AUTHORITY, parameters.getAuthority().toString());
```

Comment:
I'd personally wrap everything with try-cache, and log if there's any exception. (Mainly to cache unexpected nullpointerException or casting exception). My stance is that telemetry error shouldn't crash the app.

Reply:
I don't like plastering over bugs like this. I'd rather see unit tests and just have confidence that everything works. If we shipped an NPE that caused clients not to send telemetry how would we know?

Reply:
@rpdome  try-cache will stop the running of the following code. And we want to collect useful data as much as possible.

Reply:
@iambmelt Agree.